config/config-api/src/com/thoughtworks/go/config/PluginRoleUsersStore.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2017 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.config;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.SetMultimap;
+
+import java.util.*;
+
+import static com.google.common.collect.Multimaps.synchronizedSetMultimap;
+
+/**
+ * Singleton class to hold list of users against a PluginRole. This class will be the source of truth to find users for a given PluginRole
+ */
+
+public class PluginRoleUsersStore {
+    private final SetMultimap<PluginRoleConfig, RoleUser> roleToUsersMappings = synchronizedSetMultimap(HashMultimap.create());
+    private static PluginRoleUsersStore self;
+
+    private PluginRoleUsersStore() {
+
+    }
+
+    public static PluginRoleUsersStore instance() {
+        if (self == null) {
+            self = new PluginRoleUsersStore();
+        }
+        return self;
+    }
+
+    public void assignRole(String user, PluginRoleConfig pluginRoleConfig) {
+        roleToUsersMappings.put(pluginRoleConfig, new RoleUser(user));
+    }
+
+    public List<RoleUser> usersWithRole(PluginRoleConfig pluginRoleConfig) {
+        return new ArrayList(roleToUsersMappings.get(pluginRoleConfig));
+    }
+
+    public Set<PluginRoleConfig> pluginRoles() {
+        return roleToUsersMappings.keySet();
```

Comment:
This keyset is mutable and changes to the keyset will cause changes in the underlying map. See if you need to return a copy?